src/librbd/crypto/CryptoObjectDispatch.cc

```diff
@@ -79,7 +89,212 @@ struct C_EncryptedObjectReadRequest : public Context {
+        r = 0;
+        for (uint64_t i = 0; i < extents->size(); ++i) {
+          auto& extent = (*extents)[i];
+          auto& aligned_extent = aligned_extents[i];
+          extent.bl.substr_of(
+                  aligned_extent.bl,
+                  extent.offset - aligned_extent.offset,
+                  extent.length);
+          r += extent.length;
+        }
+      }
+      on_finish->complete(r);
+    }
+};
+
+template <typename I>
+struct C_UnalignedObjectWriteRequest : public Context {
+
+    CephContext* cct;
+    uint64_t object_off;
+    ceph::bufferlist* aligned_data;
+    ceph::bufferlist* cmp_data;
+    uint64_t* mismatch_offset;
+    io::ReadExtents extents;
+    Context* onfinish;
+    uint64_t version;
+    C_UnalignedObjectReadRequest<I>* read_req;
+    std::function<void(std::optional<uint64_t>)> write_req_sender;
+
+    C_UnalignedObjectWriteRequest(
+            I* image_ctx, CryptoInterface* crypto, uint64_t object_no,
+            uint64_t object_off, ceph::bufferlist* data,
+            ceph::bufferlist* cmp_data, uint64_t* mismatch_offset,
+            IOContext io_context, int op_flags, int write_flags,
+            std::optional<uint64_t> assert_version,
+            const ZTracer::Trace &parent_trace, int* object_dispatch_flags,
+            uint64_t* journal_tid,
+            Context* on_dispatched) : cct(image_ctx->cct),
+                                      object_off(object_off),
+                                      aligned_data(data),
+                                      cmp_data(cmp_data),
+                                      mismatch_offset(mismatch_offset),
+                                      onfinish(on_dispatched) {
+      auto aligns = crypto->get_pre_and_post_align(
+              object_off, data->length());
+      auto aligned_object_off = object_off;
+      if (aligns.first != 0) {
+        aligned_object_off -= aligns.first;
+        extents.emplace_back(aligned_object_off, aligns.first);
+      }
+      if (aligns.second != 0) {
+        extents.emplace_back(object_off + data->length(),
+                             aligns.second);
+      }
+
+      if (cmp_data != nullptr) {
+        extents.emplace_back(object_off, cmp_data->length());
+      }
+
+      read_req = new C_UnalignedObjectReadRequest<I>(
+              image_ctx, crypto, object_no, &extents, io_context,
+              0, 0, parent_trace, &version, 0, this);
+
+      write_req_sender = [image_ctx, object_no, aligned_object_off, io_context,
+                          op_flags, write_flags, assert_version, journal_tid,
+                          &parent_trace,
+                          this] (std::optional<uint64_t> version) mutable {
+        bool exclusive = write_flags & io::OBJECT_WRITE_FLAG_CREATE_EXCLUSIVE;
+        if (!version.has_value()) {
+          write_flags |= io::OBJECT_WRITE_FLAG_CREATE_EXCLUSIVE;
+          version = assert_version;
+        } else if (exclusive) {
+          onfinish->complete(-EEXIST);
```

Comment:
Who is handling these `-EEXIST`, `-EOVERFLOW`, and `-ERANGE` errors? I would think you would need to kick off a new read request on `-EEXIST` and `-ERANGE` (not sure how you get `-EOVERFLOW` besides a bug)

Reply:
I'm returning EXISTS, EOVERFLOW, ERANGE only when the user used OBJECT_WRITE_FLAG_CREATE_EXCLUSIVE or assert_version.
These types of requests are usually handled by the CORE layer, which uses RADOS api.
Since I cut down the request from going down to CORE and close the loop here on crypto layer, I checked how does RADOS handle these cases, and copied these return codes from the OSD code.
BTW same for compare_and_write which translates to rados `cmpext`, and can return -EILSEQ.

Reply:
The `-EILSEQ` is at least an expected failure from compare-and-write. Those other ones are errors that will be returned just because of the crypto-layer's internal logic. That's why I think you need to effectively (internally) handle those version or exclusive-create issues yourself and (essentially) restart the op. i.e. we don't want these errors bubbling back up the user.

Reply:
Yep, just to verify that we're aligned:
EEXISTS, EOVERFLOW, ERANGE can also be returned to the user today, and I will return them in the same cases (not caused by the crypto internal guards logic).
I will just eliminate the ERESTART and do an actual restart myself.

Reply:
I am missing something -- where would these be returned to the user today?

Reply:
... and presume you are just referring to the new `OBJECT_WRITE_FLAG_CREATE_EXCLUSIVE` and `assert_version` logic, the only user for that is this layer (right now) so this layer needs to handle those conditions to prevent them bubbling back up to the end-user.

Reply:
Yep, that was my intent.
Although crypto is the only user if these object dispatch APIs, my code should not assume that, and not implement an "always restart" logic, but rather check if the use of the API came from the Crypto layer itself, or from outside. This is what I'm trying to implement.